### ISSUE TITLE
Window rule to force windows above top layer components

### ISSUE THREAD

**Issue by Vosjedev** [hyprwm/Hyprland#4847]
### Description

Since not-so-long flameshot came with wlroots-based compositor support, on hyprland with some window rules (see [this comment](https://github.com/flameshot-org/flameshot/issues/2978#issuecomment-1931911097)). However, bars like waybar appear above the window when `layer` is set to `top`. I would like to have a window rule to make windows appear above the `top` layer. Would this be possible, and if so, would you be interested in making it?  

Additional info:  
Use `flameshot-git` from the AUR, or the source-compile the latest commit of flameshot for wlroots-support to work.

![image](https://github.com/hyprwm/Hyprland/assets/100359465/2c6ff14b-36d8-44ac-b485-ccc64c58b8f1)

**Comment by vaxerski** [comment]
fullscreen it

**Comment by gfhdhytghd** [comment]
Hello, I have the same windowsrule needs. I'm sorry that I can't send contributions directly because my programming skills are really not very good. My situation is like this. I created a script and some windowsrules to implement a "Linux Copilot" and it worked. Good, but the top is blocked by the waybar, so I also want a windowsrule that forces the window to be higher than the top layer

On my computer I could modify this GPT's window to align with the edges of other windows, but since I want to incorporate this "copilot" into my own hyprland installation script, take into account various monitor situations , I can only make it appear across the entire vertical direction, otherwise it will be uneven and it will be ugly.

To sum up, I need a rule that forces the window to be displayed higher than the top layer.

If possible, I also hope to have a rule that can set the position and size of the window by setting the distance from the edge of the window to the edge of the screen. This rule will, to a certain extent, enable the same configuration file to be used on different monitors. Better compatibility.

picture of the "Linux Copilot"
![image](https://github.com/hyprwm/Hyprland/assets/102898587/3531eb5a-1c8d-4498-b08c-3e5fed8991a1)

**Comment by gfhdhytghd** [comment]
This is My solution for screen shot

bind = $mainMod, S, exec, grim -t jpeg -q 100 -g "$(slurp)" - | swappy -f - # take a screenshot

**Comment by Vosjedev** [comment]
@gfhdhytghd Grim and slurp, or rather `grimshot`, is not really an option for me, for several reasons.

I sometimes make screenshots where timing is a problem. Slurp does not pause the screen while selecting, if something is moving it will keep moving.

I also need some other functions of flameshot, and find gimshot (even with swappy) a bit underfunctioning...

@vaxerski (sorry for pinging you) I tried fullscreening the window but it just moved the window to my smaller monitor and moved it down:
![image](https://github.com/hyprwm/Hyprland/assets/100359465/f578d5ac-ad59-4c2a-ad0f-3f074a64c68e)

**Comment by Vosjedev** [comment]
For now I solved this with the following script which just hides waybar:
```
#!/usr/bin/bash

flameshot gui &
sleep 3
killall -SIGUSR1 waybar # toggle waybar
wait # wait for flameshot to exit
killall -SIGUSR1 waybar # toggle waybar
```
you will need to ajust the sleep timing according to the speed of your computer.

I guess as I am not the only one with this problem I won't close this for now? Vaxry, you can decide if this needs to be closed or not.

**Comment by alba4k** [comment]
> fullscreen it

Add an option to fullscreen something on every screen, like you can on i3 (and I think sway too) with `fullscreen toggle global`

**Comment by shasherazi** [comment]
I also want to render my dmenu floating window over my waybar, but I can't since waybar is always on top. Is there any way of doing that currently? I tried `alterzorder` and `pin` but they didn't work.

**Comment by alba4k** [comment]
I think you can fix that by using a dmenu version that's actually made for working on wayland, and therefore can be set to use an overlay layer, rather than just using the normal one via xwayland

https://github.com/nyyManni/dmenu-wayland

I found this, not sure if you can make it appear on top of waybar rather than below it (y-wise) tho

**Comment by alba4k** [comment]
> Add an option to fullscreen something on every display, like you can on i3 (and I think sway too) with `fullscreen toggle global`

@vaxerski would this be achievable in any meaningful way?

**Comment by vaxerski** [comment]
scripts/plugins. Not planned to be a part of core. It's, for all intents and purposes, stupid.

**Comment by shasherazi** [comment]
@vaxerski what about my usecase? Can I render a window on top of wayland?

**Comment by Vosjedev** [comment]
For anyone using the window rules mentioned in that original post:
The rule name changed, so change the line
```
windowrule = nofullscreenrequest,flameshot
```
to be 
```
windowrule = fakefullscreen,flameshot
```

**Comment by alba4k** [comment]
And so will the tooltips

**Comment by vaxerski** [comment]
yea yea I will fix that one day

**Comment by alba4k** [comment]
Thanks @vaxerski for being a good boy :)

Now @Vosjedev since the original issue is fixed I guess this issue could be closed? even tho there is no real "fix"? up to you ig

**Comment by Vosjedev** [comment]
Then I'll move that to vaxerski like I already did ~3weeks ago:
> I guess as I am not the only one with this problem I won't close this for now? Vaxry, you can decide if this needs to be closed or not.

:D

I guess it (kind of)<sup>[1]</sup> works for this specific case (flameshot), but as @gfhdhytghd had some usecase for this as well, I don't know wheather this is accually resolved... But maybe I need to learn some C++ and write a plugin for this. It will probably be months before I start on that though, as I got some other big things in mind to do first. But if others (@gfhdhytghd ?) are really interested in this working, I can give it more importance on my todo list.

*[1] It is however a bit sketchy as not all UI elements get hidden, for example another overlay I added later is still in front of flameshot, and my notifications also still show up. This means it is a bit of a temporary solution to me.*

**Comment by vaxerski** [comment]
I still will not budge on the fact that accommodating for apps' laziness in doing things in a fucking shit way instead of the proper way is not what I will be interested in doing.

**Comment by alba4k** [comment]
tbh being able to fullscreen on every monitor was a neat feature to have sometimes, but sure

**Comment by Vosjedev** [comment]
> I still will not budge on the fact that accommodating for apps' laziness in doing things in a fucking shit way instead of the proper way is not what I will be interested in doing.

I am not saying you need to fix this, you already said you don't want to, and I won't push my will on people making open source software in their own time. I agree that this is indeed a flameshot issue now that I read this.

**Comment by Vosjedev** [comment]
> Or politely ask the software developers (flameshot, chrome, etc.) to provide an option to render the software at any level?

For that they would have to get the compositor to put the windows above the shell, so a wayland protocol for that would need to exist. I don't think there is any protocol for requesting normal (non-shell) windows to be above the `top` shell layer, but correct me if I'm wrong here. I also don't think they will make it 'at wayland' (the protocol developers), given the fact they are currently discussing about basic functionality : ) but I guess flameshot should be able to do the same as `slurp`, however I have no idea how hard that would be to implement. Flameshot is not very busy in supporting wayland, so also don't expect them to do that much about it...

> Maybe we should politely ask the waybar developers if they can make different parts of the bar run at different levels?

I don't think they will do that, they have better things to do I guess.

**Comment by letavocado** [comment]
Hi folks! I guess I have also new case.

### Description:
When using LunarVim as an IDE for Flutter development, the Android Emulator is set to float above other windows. However, when LunarVim is toggled to fullscreen, the Emulator window is hidden, even though it is expected to remain visible and it's okay. But how to make Emulator **always on top**?

### Expected Behavior:
The Android Emulator should remain visible on top of the fullscreen LunarVim window, maintaining its floating state.

For example, when runing wofi it's always on top.

Cheers! :sparkles: 

[
![20240809_23h56m56s_grim](https://github.com/user-attachments/assets/eac3e57a-8b2b-49b6-9e71-d3395a0816fc)
](url)

**Comment by alba4k** [comment]
Wofi is renderes as a layer, not a window

**Comment by littleblack111** [comment]
> Hello, I have the same windowsrule needs. I'm sorry that I can't send contributions directly because my programming skills are really not very good. My situation is like this. I created a script and some windowsrules to implement a "Linux Copilot" and it worked. Good, but the top is blocked by the waybar, so I also want a windowsrule that forces the window to be higher than the top layer
> 
> On my computer I could modify this GPT's window to align with the edges of other windows, but since I want to incorporate this "copilot" into my own hyprland installation script, take into account various monitor situations , I can only make it appear across the entire vertical direction, otherwise it will be uneven and it will be ugly.
> 
> To sum up, I need a rule that forces the window to be displayed higher than the top layer.
> 
> If possible, I also hope to have a rule that can set the position and size of the window by setting the distance from the edge of the window to the edge of the screen. This rule will, to a certain extent, enable the same configuration file to be used on different monitors. Better compatibility.
> 
> picture of the "Linux Copilot" ![image](https://private-user-images.githubusercontent.com/102898587/308583673-3531eb5a-1c8d-4498-b08c-3e5fed8991a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzNTc2MzQsIm5iZiI6MTcyNTM1NzMzNCwicGF0aCI6Ii8xMDI4OTg1ODcvMzA4NTgzNjczLTM1MzFlYjVhLTFjOGQtNDQ5OC1iMDhjLTNlNWZlZDg5OTFhMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwM1QwOTU1MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YmFkM2QzMTUyNjM0NDY3ZjQ5ZmI2Y2ZmYTc3MzdlZmMyODAzNDA4YzdiN2RiOTc5OWMwMzhkOGJhYzZkM2VlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.GTtBILgRa_fygyW_ToAYOMdtFMHUv-PQzL8fBP4NczQ)

how did you get that "linux copilot" seem very cool!

**Comment by gfhdhytghd** [comment]
> > Hello, I have the same windowsrule needs. I'm sorry that I can't send contributions directly because my programming skills are really not very good. My situation is like this. I created a script and some windowsrules to implement a "Linux Copilot" and it worked. Good, but the top is blocked by the waybar, so I also want a windowsrule that forces the window to be higher than the top layer你好，我有相同的窗台规则需求。很抱歉，我不能直接投稿，因为我的编程技术真的不是很好。我的情况是这样的。我创建了一个脚本和一些windows规则来实现“Linux副驾驶”，它工作了。很好，但是顶部被路径条挡住了，所以我还想要一个窗口规则，强制窗口高于顶层
> > On my computer I could modify this GPT's window to align with the edges of other windows, but since I want to incorporate this "copilot" into my own hyprland installation script, take into account various monitor situations , I can only make it appear across the entire vertical direction, otherwise it will be uneven and it will be ugly.在我的电脑上，我可以修改这个GPT的窗口，使其与其他窗口的边缘对齐，但由于我想将这个“副驾驶”合并到我自己的hyperland安装脚本中，考虑到各种显示器的情况，我只能使它出现在整个垂直方向上，否则会不均匀，会很难看。
> > To sum up, I need a rule that forces the window to be displayed higher than the top layer.总而言之，我需要一个规则来强制窗口显示在高于顶层的位置。
> > If possible, I also hope to have a rule that can set the position and size of the window by setting the distance from the edge of the window to the edge of the screen. This rule will, to a certain extent, enable the same configuration file to be used on different monitors. Better compatibility.如果可能的话，我还希望有一个规则，可以通过设置窗口边缘到屏幕边缘的距离来设置窗口的位置和大小。该规则在一定程度上允许在不同的监视器上使用相同的配置文件。更好的兼容性。
> > picture of the "Linux Copilot" “Linux Copilot”图片![image](https://private-user-images.githubusercontent.com/102898587/308583673-3531eb5a-1c8d-4498-b08c-3e5fed8991a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUzNTc2MzQsIm5iZiI6MTcyNTM1NzMzNCwicGF0aCI6Ii8xMDI4OTg1ODcvMzA4NTgzNjczLTM1MzFlYjVhLTFjOGQtNDQ5OC1iMDhjLTNlNWZlZDg5OTFhMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwM1QwOTU1MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YmFkM2QzMTUyNjM0NDY3ZjQ5ZmI2Y2ZmYTc3MzdlZmMyODAzNDA4YzdiN2RiOTc5OWMwMzhkOGJhYzZkM2VlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.GTtBILgRa_fygyW_ToAYOMdtFMHUv-PQzL8fBP4NczQ)
> 
> how did you get that "linux copilot" seem very cool!你怎么知道“linux副驾驶”看起来很酷!
script:
https://github.com/gfhdhytghd/hyprland-setup/blob/main/HyprV/hypr/scripts/sidechat.sh
winrule to make it work properly:
(A little bit of change will be needed to make it fit your desktop style)
```
#chat drop down
exec-once= touch /tmp/chattoggle
windowrule=float,chrome-chat.openai.com__-Default
windowrule=size 25% 93.2%,chrome-chat.openai.com__-Default
windowrule=move 74.28% 5.63%,chrome-chat.openai.com__-Default
```
you need a key bind to exclute this script in order to make it work,like this.
```
bind = $mainMod, SPACE, exec, bash $HOME/.config/hypr/scripts/sidechat.sh
```
I recommand you to set uo fusuma in order to use touchpad to call it if you are using a laptop.

**Comment by littleblack111** [comment]
nice. nah im on desktop with mouse..

but cool script man!

**Comment by Blackilykat** [comment]
I have a similar problem where I have multiple screens and need flameshot to be on top of all other windows. I got a workaround that allowed me to do so for floating and tiling windows, but it still stays under fullscreen windows.

These are my current flameshot window rules:
```
windowrulev2 = animation fade, class:(flameshot)
windowrulev2 = norounding, class:(flameshot)
windowrulev2 = noborder, class:(flameshot)
windowrulev2 = fullscreenstate 0 0, class:(flameshot)
windowrulev2 = float, class:(flameshot)
windowrulev2 = pin, class:(flameshot)
windowrulev2 = monitor DP-1, class:(flameshot)
windowrulev2 = move 0 0, class:(flameshot)
windowrulev2 = size 200% 100%, class:(flameshot)
```

**Comment by FedotCompot** [comment]
Stumbled upon this thread by chance, been migrating to hyprland from i3 this weekend, flameshot has always been a killer feature.

I ended up with a similar config to @Blackilykat :
```
windowrulev2 = stayfocused, class:flameshot, title:flameshot
windowrulev2 = norounding, class:flameshot, title:flameshot
windowrulev2 = noborder, class:flameshot, title:flameshot
windowrulev2 = float, class:flameshot
windowrulev2 = pin, class:flameshot, title:flameshot
windowrulev2 = move 0 0, class:flameshot, title:flameshot
```

Differences are
- `stayfocused` rule to not lose focus when the cursor is over the waybar
- double selector so other flameshot windows are not matched (Settings, Launcher, etc...) except for floating and pin
- `size`, `monitor`, `fullscreenstate` are missing, because for me it worked without setting them

Have been happy with it so far, hope someone finds this useful in their adventure to the waylands

**Comment by vaxerski** [comment]
Hello there!

This issue has been closed, as we are moving from issues to discussions. More about this can be found [here](https://wiki.hyprland.org/Contributing-and-Debugging/Issue-Guidelines/) and in #9854.

Firstly, please make sure this issue is still relevant in latest Hyprland. If so, we ask you to open a [discussion](https://github.com/hyprwm/Hyprland/discussions) (please read the discussion guidelines first, in the pinned post)

Thank you for your understanding!
